### DISCUSSION TITLE
Redirect in middleware when request is a server action fails with "failed to forward action response"

### DISCUSSION THREAD

**Discussion by ehis** [vercel/next.js#64993] [Help] [answered]
Running: `NextJS v14.2.3` 
Node: `v20.5.1`

Here's the [permalink](https://github.com/vercel/next.js/blob/e078fd1715254c5c64013b79dbeefcb3a42bc9ac/packages/next/src/server/app-render/action-handler.ts#L230) to the `console.error` in the `action-handler`.

## Stack Trace on Vercel
```
failed to forward action response TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11731:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async r_ (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:4134)
    at async rP (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:7924)
    at async r5 (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:1139)
    at async es (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:16:26320)
    at async ei.responseCache.get.routeKind (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:1026)
    at async r3.renderToResponseWithComponentsImpl (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:508)
    at async r3.renderPageComponent (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:5028)
    at async r3.renderToResponseImpl (/var/task/node_modules/.pnpm/next@14.2.3_@babel+core@7.24.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:5615) {
  cause: Error
      at makeNetworkError (node:internal/deps/undici/undici:5002:35)
      at httpRedirectFetch (node:internal/deps/undici/undici:9994:16)
      at httpFetch (node:internal/deps/undici/undici:9954:28)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async schemeFetch (node:internal/deps/undici/undici:9854:18)
      at async node:internal/deps/undici/undici:9727:20
      at async mainFetch (node:internal/deps/undici/undici:9717:20)
}
```

**Comment by ehis** [comment]
hi @ztanner In our application, `createForwardedActionResponse` fetch request is failing when the `middleware` intercepts a server action request and based on some application logic (eg. invalid session) redirects the user to another page.

https://github.com/vercel/next.js/pull/64227

Example of condition and redirect in `middleware`.

```js

const isMember = false

if (!isMember) {
   return NextResponse.redirect(new URL('/plans', request.url));
}


```

Hoping we can get a patch soon!

**Comment by ztanner** [comment]
Hi @ehis, thank you for the report! This should be fixed by https://github.com/vercel/next.js/pull/65097 which is available in `v14.3.0-canary.27`. Let me know if you're still seeing this behavior on/after this version.

**Reply by ehis** [reply]
hi @ztanner. Thanks for the response! That PR resolves the fetch request. However, noticed the response is now `undefined` which causes a render error when reading values from the server action response.


Here's a [codesandbox](https://codesandbox.io/p/devbox/wizardly-gwen-zvwj28?file=%2Fapp%2Fpage.tsx%3A24%2C26&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clvlksd6500073b8sr4ycmb8l%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clvlksd6500023b8su0ogseto%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clvlksd6500043b8s8pns6p79%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clvlksd6500063b8sy5ba8l4l%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clvlksd6500023b8su0ogseto%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clvlksd6400013b8sfyxjbu2m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clvlkwzw100023b6opq84uq7z%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A24%252C%2522startColumn%2522%253A26%252C%2522endLineNumber%2522%253A24%252C%2522endColumn%2522%253A26%257D%255D%252C%2522filepath%2522%253A%2522%252Fapp%252Fpage.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clvlksd6500023b8su0ogseto%2522%252C%2522activeTabId%2522%253A%2522clvlkwzw100023b6opq84uq7z%2522%257D%252C%2522clvlksd6500063b8sy5ba8l4l%2522%253A%257B%2522id%2522%253A%2522clvlksd6500063b8sy5ba8l4l%2522%252C%2522activeTabId%2522%253A%2522clvllbgf100xw3b6o469c082d%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clvlksd6500053b8sibc2l74j%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252Fplans%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clvllbgf100xw3b6o469c082d%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252Fbuy%2522%257D%255D%257D%252C%2522clvlksd6500043b8s8pns6p79%2522%253A%257B%2522id%2522%253A%2522clvlksd6500043b8s8pns6p79%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clvll5yw200m43b6oxcjikjqf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clvll5yzf0018dgfm0u4f46yy%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clvll5yw200m43b6oxcjikjqf%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) with the scenario.


To repro in codesandox navigate to /buy and click the "purchase me"

If you take a look at the console logs you'll see response comes back as undefined

**Reply by ehis** [reply]
### more observations

- seems like the middleware is re-running the `/buy` page
- should probably be `GET /plans` and not `POST /plans` since it's not a server action

```
request.url /buy
redirecting to /plans
request.url /plans
request.url /buy
redirecting to /plans
 POST /plans 200 in 51ms
```

**Reply by ztanner** [reply]
Hi @ehis -- could you help me understand a bit better what you're trying to do with the redirect? A couple things I noticed:

- It's importing `getIsServerAction`, which is an internal Next.js util that I wouldn't recommend relying on, as it could change/be removed at any time. 
- Your middleware is redirecting the POST request to `/buy` to `/plans`. This is because the `NextResponse.redirect` is explicitly looking for server action requests and routing them to a different URL. Your `/plans` page doesn't have any server actions defined on it. Server actions work by creating an implicit `POST` handler on the page(s) that import/define the action. So in this case, a `POST` handler is created on your `/buy` route.
- Next.js will attempt to forward your action to a worker that can handle it (in this case, the action that's part of `/buy`). But your redirect rule is going intercept that one too. That means your action can never run, because the redirect rule in middleware isn't letting it find the handler.

**Reply by ehis** [reply]
@ztanner Thanks for the callout on `getIsServerAction` and only using to demonstrate behavior - not going to use in production code.

Correct, as presumed the end goal is for the middleware to intercept any server action triggered on `/buy` (including the forwarded action) and force redirect to the `/plans` page.

A practical use case for this flow could be the middleware intercepting an action and redirecting to the `/login` page because of an expired JWT.

If there's another recommended NextJS approach to doing the equivalent would love any suggestions 🙏🏾

**Reply by ztanner** [reply]
Hey @ehis --

I see, that makes sense! For the use-case you shared, some ideas/recommendations:

- Ensure your server action is performing the hard auth check, rather than solely relying on middleware ([ex](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#authentication-and-authorization)). 
- If the user isn't logged in, your server action could trigger a `redirect` to the plans page, eg:

```tsx
export const purchaseMe = async () => {
  const isLoggedIn = await verifyAuth()

  if (!isLoggedIn) {
    redirect('/plans')
  }

  return {
    foo: 'bar',
  }
}
```
- If you wanted to generalize this behavior, you could abstract it into a helper that's used on your non-public actions. e.g., `await redirectIfUnauthorized()`)

**Reply by ehis** [reply]
@ztanner I see! If we perform the hard auth check in our server actions could we then exclude them from the middleware? 

[eg](https://nextjs.org/docs/app/building-your-application/routing/middleware).

```ts

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    {
      source:
        '/((?!api|_next/static|_next/image|media|fonts|favicon.ico|favicon.png).*)',
      missing: [
        // Exclude Server Actions
        { type: 'header', key: 'next-action' },
      ],
    },
  ],
};

```

**Reply by ztanner** [reply, marked as answer]
@ehis, yep, no need to handle in middleware! You can think of it similar to an API route where you'd have the auth check at the top of the request handler.

**Reply by ehis** [reply]
thanks so much @ztanner and really appreciate the response to the help discussion!!

**Reply by EnzoPV25** [reply]
hey, I had a similar issue, and I could solve it by adding the missing field. However I don't really understand why excluding server actions does the trick, and even more, how is this line excluding server actions ?

My use case was an auth process with jwt token and refresh token, that would redirect to login when the session was expired

**Reply by harsha-at-magicshake** [reply]
@Enzo-PVsyst Same in my case.
My use case is something like this:

middleware.ts

```ts
const testAuth = await isAuthenticated();
  console.log(testAuth, "in middleware");

  if (testAuth) {
    return NextResponse.next();
  } else {
    return NextResponse.redirect(new URL("/login", request.url));
  }
 }
// matcher config that excludes login and some other essentials
```

login/page.tsx
```tsx
"use client";

import { useRouter } from "next/navigation";
import { makeAPICall } from "../utils";

export default function Home() {
  const router = useRouter();

  const handleSet = async (value: number) => {
    await makeAPICall(value); // added this implementation at the end. (utils.ts)
    // redirect("/v4") // commented because it's throwing NEXT_REDIRECT error.
    router.replace("/v4");
  };

  return (
    <main className="space-x-4">
      <button onClick={async () => await handleSet(1)}>Set Cookie</button>
    </main>
  );
}
```

v4/page.tsx
```tsx
"use client";

import { useRouter } from "next/navigation";
import { makeAPICall, onSubmit } from "../utils";
import clearCachesByServerAction from "@/customCacheClear";

export default function Home() {
  const router = useRouter();

  const handleSet = async (value: number) => {
    await makeAPICall(value);
    clearCachesByServerAction("/");
    if (value == -1) {
      router.replace("/login");
    }
  };

  return (
    <main className="space-x-4">
      <button onClick={async () => await handleSet(-1)}>Logout</button>
      <button onClick={async () => await handleSet(0)}>Get Cookie</button>
      <button onClick={async () => await handleSet(1)}>Set Cookie</button>
      <button onClick={async () => await handleSet(2)}>Refresh Cookie</button>
    </main>
  );
}
```

utils.ts
```ts
"use server"

export const isAuthenticated = async () => {
  const value = await getCookie();
  if (value) {
    return true;
  } else {
    return false;
  }
};

export const setCookie = async () => {
  cookies().set("client", "settingCookie");
};

export const deleteCookie = async () => {
  cookies().delete("client");
};

export const setNewCookie = async () => {
  cookies().set("client", "settingNewCookie");
};

export const getCookie = async () => {
  const value = cookies().get("client");
  console.log(value?.value ?? "Not found!");
  return value?.value;
};
```

clearCacheByServerAction.ts
```ts
"use server"
import { revalidatePath } from "next/cache"

const clearCachesByServerAction = async (path: any) => {
try {
  if (path) {revalidatePath(path)} 
else {
  revalidatePath("/")
  revalidatePath("/[lang]")
}
} catch (error) {
  console.error("clearCachesByServerAction=> ", error)
}
}
export default clearCachesByServerAction
```


## Use case:
When I click on logout, the cookies gets deleted, but still, apparently, in the middleware, the log shows
`true` -> which states user is authenticated

The intended action was to navigate to `/login` which wouldn't happen without putting `clearCacheByServerAction`
As per my understanding, when I clear cache, the middleware is hit again, then it shows false, hence it gets redirected.
In this scenario, I am getting the error.

Now my queries are:
Is it okay to get it? 
Is it intended? 
Can I go ahead to production without handling it?
Why `redirect("/")` getting NEXT_REDIRECT error?
Can I include checking the token validity logic in `middleware.ts` and handle refresh token there? Or is it heavy to do so?
Assuming `makeApiCall` is a helper which calls all my routes, from all the screens, can I use it to intercept the responses, (I handled the token expiry with custom code) so that I can check it and do another API call to refresh token, set new cookies, then perform the actual api call?

@ztanner 

Apologies in advance if the queries seem to be silly, since I did not find any proper resource for handling the cookies from route.ts, middleware.ts etc.

**Reply by Daniel-Ash** [reply]
Hey @ztanner - so to confirm, are you saying that you can't use middleware for auth redirects when using server actions? It would be great if this was supported, or added to the middleware docs as a limitation as it can lead to bugs that are tricky to pin down!

**Reply by smozely** [reply]
Yeah this totally should be documented somewhere more obvious. The [Middleware Logic](https://nextjs.org/docs/app/building-your-application/authentication#middleware-logic) section in the docs makes no mention of this, and the example code is misleading. And the [Protecting Server Actions](https://nextjs.org/docs/app/building-your-application/authentication#protecting-server-actions) sections says to make sure you protect server actions, but it doesn't anywhere mention that the middleware can't be used. 


> yep, no need to handle in middleware! 

Makes it sound like a choice, when in fact its "you can't do this from the middleware"

**Reply by kennataddese6** [reply]
> @ehis, yep, no need to handle in middleware! You can think of it similar to an API route where you'd have the auth check at the top of the request handler.

Having related issue with middleware and server action
 Problem Summary
1. I have a form using useActionState + server action 

2. On submission, middleware runs.

3. If auth() fails in middleware, it does NextResponse.redirect('/login').

4. That sends a 301 response, not a response expected by useActionState.

So, your client doesn't receive { success, errorMessage } — it gets a redirect response instead → resulting in undefined state and no redirect login.

What makes it tricky. The request never reaches server actions so using middleware for auth pages during page naviation works to redirect the user to logn but not during server action

**Reply by ibai-atienza** [reply]
@ztanner with the last enablement of the Nodejs runtime for the middleware is still not recommended the usage of the middleware to do some auth checks (token validity, etc)? The code added to the released features in the version 15.5.5 just do that verification of the token https://nextjs.org/blog/next-15-5#nodejs-middleware-stable

Any info that could clarify that point please? i was testing this approach and the redirect didn't work if the middleware detects an invalid token into the middleware if it was reached by a server function call.

Thanks for the response in advance!

**Comment by VivekGupta137** [comment]
I came across to this problem as well, IMO, having the redirects at middleware level made more sense, instead of duplicating same logic in more than one place ( one at middle ware and others at each server action ). 
I feel that if I made the middleware that captures all the requests for a particular page ( including all the server actions calls ), and that route is only for authenticated users, then it makes less sense to have separate duplicate logic in each of the server actions for that page.

My dependencies:
"dependencies": {
    "react": "^18",
    "react-dom": "^18",
    "next": "14.2.14"
  }
  
in the `middleware.ts` file:
`import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";


export function middleware(request: NextRequest) {
    const token = request.cookies.get('access_token');

    const isAuthenticated = !token? false : true;
    if (!isAuthenticated) {
        const response = NextResponse.redirect(new URL('/?error=not-authenticated', request.url));

        return response;
    }
    return NextResponse.next();
}

export const config = {
  matcher:[ "/gqlclientcheck/:path*", "/gqlservercheck/:path*"],
};`

**Comment by ollebergkvist** [comment]
@ztanner 

We're in October now, and the docs are still not updated, it should clearly reflect this conflict between server actions and running auth in middleware. Would probably save other devs from running into this.. 🙏 

Thanks to finding this closed thread we were able to solve a major headache in our auth flow.

We disabled middleware for server actions by checking this condition in our conditional statement:
const isServerAction = request.headers.get('next-action')

And added explicit checks in each action instead.

**Comment by J4v4Scr1pt** [comment]
Thanks for this thread and the "work arrounds" I was pulling my hear on this one 😅

**Comment by axten** [comment]
I solved redirection of server actions in middleware with setting the corresponding header:

```
// middleware.ts
    ...
      const hasSession = request.auth && !request.auth.isExpired;
      const isAction = request.method === 'POST' && request.headers.has('Next-Action');
    
    // since middleware cannot redirect server actions directly, the corresponding header gets set
    if (!hasSession && isAction) {
      response.headers.set('x-action-redirect', `/login?callbackUrl=${encodeURIComponent(request.nextUrl.toString())}`);
     return response;
    }
    ...
```

**Reply by ManuContrerasSV** [reply]
This worked for me. Thanks!

**Reply by Seb-klay** [reply]
For those using *Zod validation* with Server action, I updated this work around : 

```ts
  const isAction = request.method === 'POST' && request.headers.has('Next-Action');
  // since middleware cannot redirect server actions directly, the corresponding header gets set
  if (!hasSession&& isAction) {
    return ;
  }
```
Then you can safely show the errors like this : 
```ts
    const response = await getUserWithEmail(userToLogin);
    if (!response?.ok) {
      return {
        errors: {
          ...emptyLoginErrors,
          email: ["No user could be found."],
        },
      };
    }
```
Hope this helps.

**Comment by LikeDreamwalker** [comment]
There are almost nothing in document mentioned that middleware can't control server action directly.
I know writing documents is a hard job but if there is just one saying, will be better than nothing. I even exclude if SSG page can't handle 307 redirect.

**Comment by Rrobinvip** [comment]
@axten has a great proposal. However, I've noticed that the redirect with this header doesn't occur immediately. This could be due to our different setups. In my setup, I return an empty response and everything works fine.

Also, using `nextUrl.toString()` will provide the full path. Most likely, it will be `https://localhost:8080` because the server action is executed on the deployment.

Here's a revised code:

```
// Change return url to pathname and search

const response = new NextResponse();
response.headers.set('x-action-redirect', `/login?returnUrl=${encodeURIComponent(request.nextUrl.pathname + request.nextUrl.search)}`);

// Clean up any session cookies if necessary

response.cookies.delete("sessionToken");
return response;

```
This code changes the return URL to pathname and search, and cleans up any session cookies if necessary.

**Comment by kennataddese6** [comment]
Having related issue with middleware and server action
 Problem Summary
1. I have a form using useActionState + server action 

2. On submission, middleware runs.

3. If auth() fails in middleware, it does NextResponse.redirect('/login').

4. That sends a 301 response, not a response expected by useActionState.

So, your client doesn't receive { success, errorMessage } — it gets a redirect response instead → resulting in undefined state and no redirect login.

What makes it tricky. The request never reaches server actions so using middleware for auth pages during page naviation works to redirect the user to logn but not during server action.

**Comment by kennataddese6** [comment]
Yes. I believe this is serious

Kind Regards,
[image: image]
Kenna Taddese
Software Developer
Phone. ‪+1 (530) 712-3425‬‬
Web. Babure.rf.gd
Address. Level 9, Abraham Street, WA 6029
[image: facebook] <http://www.facebook.com/> [image: twitter]
<http://www.twitter.com/> [image: linkedin] <http://www.instagram.com/> [image:
google-plus] ***@***.***/>
Get your own signature
<https://gimm.io/en_US/email-signature-generator?utm_source=sent-emails&utm_medium=email&utm_campaign=get-your-own-signature>
